clean-modular-code/checks-conditionals/python-function-checks.md

```diff
@@ -11,79 +11,112 @@ kernelspec:
-
-A try/except block looks like this:
+When dealing with messy or unpredictable data, it’s important to ensure your code
+can handle errors early and gracefully. This approach is often called the **fail fast** approach. 
 
-```python
-try:
-    # code that might cause an error
-except SomeError:
-    # what to do if there's an error
-```
+Your goal is to identify detect and data processing or workflow problems immediately when they occur, rather than allowing
+them to propagate through your code. This approach saves time and makes
+debugging easier, as it provides clearer, more useful error outputs (known as stack traces).
 
 ```{code-cell} ipython3
 ---
 editable: true
 slideshow:
   slide_type: ''
+tags: [raises-exception]
 ---
-def convert_to_int(value):
-    try:
-        return int(value)
-    except ValueError:
-        print("Oops i can't process this so I will fail gracefully.")
-        return None  # or some default value
+# Open a file (but it doesn't exist
+def read_file(file_path):
+    with open(file_path, 'r') as file:
+        data = file.read()
+    return data
+
+file_data = read_file("nonexistent_file.txt")
 ```
 
++++ {"editable": true, "slideshow": {"slide_type": ""}}
+
+In the example below, you use a [conditional statement](python-conditionals) to check to see if the file exists; if it doesn't, then it returns None. In this case the code will fail quietly and the user doesn't understand that there is an error. 
+
 ```{code-cell} ipython3
 ---
 editable: true
 slideshow:
   slide_type: ''
 ---
-convert_to_int("123")
+import os
+
+def read_file(file_path):
+    if os.path.exists(file_path):
+        with open(file_path, 'r') as file:
+            data = file.read()
+        return data
+    else:
+        return None  # Doesn't fail immediately, just returns None
+
+# No error raised, even though the file doesn't exist
+file_data = read_file("nonexistent_file.txt")
 ```
 
++++ {"editable": true, "slideshow": {"slide_type": ""}}
+
+This code example below is better than the examples above for three reasons:
+
+1. It's pythonic: it asks for forgiveness later by using a try/except
+2. it fails quickly - as soon as it tries to open the file. The code won't continue to run after this step fails.
+3. it raises a clean, useful error that the user can undersatnd
+is an example of Pythonic code. Here, a file is read in and
+
+The code anticipates what will happen if it can't find the file. Here you can raise a `FileNotFoundError` and provide a useful message to the user. 
+
 ```{code-cell} ipython3
 ---
 editable: true
 slideshow:
   slide_type: ''
+tags: [raises-exception]
 ---
-convert_to_int("abc") 
+def read_file(file_path):
+    try:
+        with open(file_path, 'r') as file:
+            data = file.read()
+        return data
+    except FileNotFoundError as e:
```

Comment:
```suggestion
    except FileNotFoundError:
```

`as <var>` is an optional part of the except statement, and should be left out unless `e` is actually needed in the block